### PR TITLE
Add feature to cast from 'borrowed' to 'unmanaged'

### LINKED ISSUE #12730: Inability to cast from `borrowed C` to `unmanaged C`
### Summary of Problem

Given a borrow that I know borrows from an unmanaged class, it would be beneficial to be able to cast it from `borrowed C` (say) to `unmanaged C`.  Today this doesn't work, yet we can cast it to simply `unmanaged`.  This feels like a non-orthogonality and therefore a bug (or, arguably, a feature request).

### Steps to Reproduce
**Associated Future Test(s):**
[`test/classes/delete-free/unmanaged/cast-to-unmanaged-to-free.chpl`](
           https://github.com/chapel-lang/chapel/blob/master/test/classes/delete-free/unmanaged/cast-to-unmanaged-to-free.chpl
          ) #12683

### Configuration Information

- Output of `chpl --version`: `chpl version 1.20.0 pre-release (4d906d1)`

### PR DESCRIPTION
**Pull request by dgarvit**
Resolves #12730.

### CODE REVIEW

**Comment by bradcray** [comment]
Assuming this fixes the future in question, the PR should remove its .future and .bad files.  There's a chance you may also want to add a new test to try a greater number of scenarios if you think that would be of value to lock the feature in (I'm saying that as a general principle, not because I see an obvious need for it).

**Comment by dgarvit** [comment]
Removed the `.future` and `.bad` files. Regarding test scenarios, I can't think of any particular ones. Can you suggest some cases?

**Comment by bradcray** [comment]
> Regarding test scenarios, I can't think of any particular ones. Can you suggest some cases?

I was imagining just illustrating different code patterns that cast from borrowed to unmanaged before freeing apart from the one in the original future.

@mppf: Would you review this PR?

File: modules/internal/ChapelBase.chpl [hunk 1]
```diff
@@ -1272,6 +1272,11 @@ module ChapelBase {
     return if x != nil then __primitive("dynamic_cast", t, casttmp) else __primitive("cast", t, nil);
   }
 
+  // this version handles borrow -> unmanaged
+  inline proc _cast(type t:unmanaged, x:borrowed) {
+    return if x != nil then __primitive("dynamic_cast", t, x) else __primitive("cast", t, nil);
```
Comment (mppf) [inline modules/internal/ChapelBase.chpl]: This overload implements a dynamic cast so would need a where clause. I expect that we actually need 2 new _cast functions. One is for the case in which the target type is a parent class of the source value's type. The other is when the target type is a child class / subclass of the source value's type.

**Comment by mppf** [comment]
@dgarvit - PR #12999 will fix this (it seemed you weren't doing more with it and I needed to solve the problem for other reasons). Let's close this PR, thanks.

**Comment by ronawho** [comment]
Closing based on https://github.com/chapel-lang/chapel/pull/12750#issuecomment-491435822

**Pull request closed without merging**